transactions/transactions_service.go

```diff
@@ -985,6 +1095,120 @@ func (svc *transactionsService) getAppIdFromCustomRecords(customRecords []lnclie
 	return nil
 }
 
+func (svc *transactionsService) SettleHoldInvoice(ctx context.Context, preimage string, lnClient lnclient.LNClient) (*Transaction, error) {
+	if len(preimage) != 64 {
+		return nil, errors.New("invalid preimage format")
+	}
+	preimageBytes, err := hex.DecodeString(preimage)
+	if err != nil {
+		return nil, fmt.Errorf("invalid preimage hex: %w", err)
+	}
+
+	paymentHashBytes := sha256.Sum256(preimageBytes)
+	paymentHash := hex.EncodeToString(paymentHashBytes[:])
+
+	var dbTransaction db.Transaction
+	result := svc.db.Limit(1).Find(&dbTransaction, &db.Transaction{
+		Type:        constants.TRANSACTION_TYPE_INCOMING,
+		State:       constants.TRANSACTION_STATE_ACCEPTED,
+		PaymentHash: paymentHash,
+	})
+
+	if result.RowsAffected == 0 {
+		logger.Logger.WithField("paymentHash", paymentHash).Error("Failed to find accepted hold invoice")
+		return nil, errors.New("failed to find accepted hold invoice")
+	}
+
+	err = lnClient.SettleHoldInvoice(ctx, preimage)
+	if err != nil {
+		logger.Logger.WithFields(logrus.Fields{
+			"preimage": preimage,
+		}).WithError(err).Error("Failed to settle hold invoice via LN client")
+		// Don't mark DB as failed here, as the settle might succeed later or might have already succeeded.
+		return nil, err
+	}
+
+	var settledTransaction *db.Transaction
+	err = svc.db.Transaction(func(tx *gorm.DB) error {
+		var err error
+		settledTransaction, err = svc.markTransactionSettled(tx, &dbTransaction, preimage, 0, false) // Assuming not self-payment for now
+		return err
+	})
+
+	if err != nil {
+		logger.Logger.WithFields(logrus.Fields{
+			"paymentHash": paymentHash,
+			"preimage":    preimage,
+		}).WithError(err).Error("Failed DB transaction while settling hold invoice")
+		return nil, err
+	}
+
+	if settledTransaction == nil {
+		return nil, fmt.Errorf("settled transaction is nil despite no error")
```

Comment:
we should avoid code like this. If there is no error the method must return what we expect.